File: Django/api/cache.py
import threading

from .models import Url


class UrlCache:
    def __init__(self):
        self._lock = threading.Lock()
        self._urls: dict[str, dict] = {}

    def load(self):
        with self._lock:
            self._urls.clear()
            for u in Url.objects.values("id", "original", "accesses"):
                self._urls[u["id"]] = {
                    "original": u["original"],
                    "accesses": u["accesses"],
                }

    def get_all(self) -> dict[str, dict]:
        with self._lock:
            return dict(self._urls)

    def get(self, id: str) -> dict | None:
        with self._lock:
            return self._urls.get(id)

    def exists(self, id: str) -> bool:
        with self._lock:
            return id in self._urls

    def set(self, id: str, info: dict):
        with self._lock:
            self._urls[id] = info

    def delete(self, id: str):
        with self._lock:
            self._urls.pop(id, None)

    def increment_accesses(self, id: str) -> int:
        with self._lock:
            if id in self._urls:
                self._urls[id]["accesses"] += 1
                return self._urls[id]["accesses"]
            return 0


url_cache = UrlCache()

File: Node.js/src/cache.js
class UrlCache {
  constructor() {
    this._urls = {};
  }

  load(urls) {
    this._urls = {};
    for (const u of urls) {
      this._urls[u.id] = { original: u.original, accesses: Number(u.accesses) };
    }
  }

  getAll() {
    const result = {};
    for (const [key, val] of Object.entries(this._urls)) {
      result[key] = { ...val };
    }
    return result;
  }

  get(id) {
    return this._urls[id] ? { ...this._urls[id] } : null;
  }

  exists(id) {
    return id in this._urls;
  }

  set(id, info) {
    this._urls[id] = info;
  }

  delete(id) {
    delete this._urls[id];
  }

  incrementAccesses(id) {
    if (this._urls[id]) {
      this._urls[id].accesses += 1;
      return this._urls[id].accesses;
    }
    return 0;
  }
}

module.exports = new UrlCache();
